NHP fMRIVolume

TakJim opened the pull request:
This PR applies fMRIVolume to NHP data.

Some additional modifications are still required. The parts that I have not been able to address yet are listed below.

### TODO
- [x] In GenericfMRIVolumeProcessingPipelineBatchNHP.sh, accept arguments via opts_AddMandatory / opts_AddOptional
- [x] In BatchNHP.sh, reorganize the parameters defined in hcppipe_conf.txt and allow users to specify them directly within BatchNHP.sh
- [x] Apply the IsLongitudinal=TRUE processing logic to PipelineBatchNHP.sh

TakJim committed:
Add NHP-specific files

TakJim committed:
Add support for NHP-specific brain scaling in MotionCorrection.sh

TakJim committed:
add adjusting reference image selection for NHP species

TakJim committed:
add NHP-specific batch script

glasserm reviewed (changes requested):
My overall comment is that I think we should properly merge the NHP changes in to the fMRIVolume pipeline rather than bifurcate the pipelines.  There is a lot of reused code between fMRIVolume and the BBR code and these are not really a completely separate process like they were for FreeSurferNHP.  On initial review the changes to one step resampling and motion correction looked okay.

TakJim committed:
Remove unnecessary post-processing after fMRIVolume execution

TakJim committed:
Enhance "PipelineNHP.sh" with comprehensive option parsing as same as "Pipeline.sh" .

TakJim commented:
@glasserm 
https://github.com/Washington-University/HCPpipelines/pull/372#pullrequestreview-3578564560

Thank you for the comment — that makes sense.

To confirm, should
`fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbasedNHP.sh`
also be combined into the main `DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh`, rather than being kept as an NHP-specific variant?

After discussing with @takuya-hayashi, we felt that this script includes quite a lot of NHP-specific logic.
I’d appreciate your thoughts on whether it should still be merged or kept separate.

glasserm commented:
> @glasserm [#372 (review)](https://github.com/Washington-University/HCPpipelines/pull/372#pullrequestreview-3578564560)
> 
> Thank you for the comment — that makes sense.
> 
> To confirm, should `fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbasedNHP.sh` also be combined into the main `DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh`, rather than being kept as an NHP-specific variant?
> 
> After discussing with @takuya-hayashi, we felt that this script includes quite a lot of NHP-specific logic. I’d appreciate your thoughts on whether it should still be merged or kept separate.

Yes please.

TakJim committed:
Adjust option handling to be same as human

TakJim committed:
Add additional options and processes for NHP.

TakJim committed:
Add additional options and processes for NHP.

TakJim committed:
remove NHP-specific version pipeline

TakJim commented on fMRIVolume/GenericfMRIVolumeProcessingPipelineNHP.sh:
```diff
@@ -0,0 +1,967 @@
+#!/bin/bash
```
remove NHP.sh in f9fd9461584760415ab0006e7c46bbb8145d9508.

TakJim committed:
Fix indentation and comment

TakJim committed:
remove NHP-specific version DistortionCorrection

glasserm commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,348 @@
+#!/bin/bash 
```
This batch script is quite different from the others.  It also seems to require arguments, whereas typically a batch script is mean to be edited appropriately for each dataset and then directly run.  SetSpecies.sh does not appear to be used here. It should behave more like PreFreeSurfer I think.

glasserm commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,348 @@
+
+     #${FSLDIR}/bin/fsl_sub $QUEUE -l $StudyFolder/$Subject/logs \
+     if [ $DryRun = "FALSE" ] ; then
+     ${HCPPIPEDIR}/fMRIVolume/GenericfMRIVolumeProcessingPipelineNHP.sh \
```
Needs to call the non NHP version.

glasserm commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,348 @@
+      if [ $DryRun = FALSE ] ; then
+       CMD="set --"
+      else
+       CMD="${HCPPIPEDIR}/fMRIVolume/GenericfMRIVolumeProcessingPipelineNHP.sh"
```
Non-NHP version.

TakJim replied:
Fixed it in 80c6e1ccf415541567aea90f01b71f3ccf5f411c.

glasserm commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,348 @@
+
+ if   [[ "${RunMode}" = 4 ]] ; then
+
+    ${HCPPIPEDIR}/fMRIVolume/GenericfMRIVolumeMotionCorrectXRunsNHP.sh $StudyFolder $Subject $AllfMRINames -x
```
Agree.  The Batch script should launch a pipeline and terminate.  If this cross-run motion correction helper function is needed for some NHP situations, it probably should have its own Batch script and run after fMRIVolume.

TakJim replied:
Thanks — I’ve added the file to the PR.
Regarding the Batch script organization, I’ll discuss with @takuya-hayashi.
bed9d933a5ca445b8e9ed07ba4736b159c15a3b0

glasserm commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,348 @@
+[ "$2" = "" ] && Usage
+
+if [ "$SPECIES" = "" ] ; then echo "ERROR: please export SPECIES first to any of Macaque, MacaqueCyno, MacaqueRhesus, Marmoset, NightMonkey, Chimp, Human"; exit 1;fi
+if [ "$HCPPIPEDIR" = "" ] ; then echo "ERROR please export HCPPIPEDIR before running this script"; exit 1; fi
```
We don't want to rely on environment variables--we should set things up like PreFreeSurfer.

glasserm commented on fMRIVolume/scripts/OneStepResampling.sh:
```diff
@@ -162,13 +162,18 @@ NumFrames=$(${FSLDIR}/bin/fslval ${InputfMRI} dim4)
+    fi
+else # TH - e.g. NHP
+    ${FSLDIR}/bin/flirt -interp spline -in ${T1wImage} -ref ${T1wImage} -applyisoxfm $FinalfMRIResolution -out ${WD}/${T1wImageFile}.${FinalfMRIResolution}
+    ResampRefIm=${WD}/${T1wImageFile}.${FinalfMRIResolution}
```
Invert to use ! Human as elsewhere?

TakJim replied:
Fixed it in 33973fa4. Now uses `if [[ $SPECIES != Human ]]` pattern.

TakJim committed:
Fix order of parameters.

TakJim commented on fMRIVolume/scripts/MotionCorrection.sh:
```diff
@@ -26,7 +26,8 @@ OutputMotionMatrixFolder="\$6"
 fMRIReferenceReg="\$9"
-
+SPECIES="\$10"
+BrainScaleFactor="\$11"
```
Fixt it in 9eb7df9482f7252cbff64c77c56f9631ef371892.

TakJim committed:
Add NHP-specific file

TakJim committed:
Fix: run the unified Human/NHP pipeline

TakJim commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,348 @@
+   "d" ) DryRun="TRUE";;
+    * )  usage_exit;;
+ esac
+done;
```
I’ll discuss with @takuya-hayashi and rename it to something clearer.

glasserm commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -133,6 +139,25 @@ opts_AddOptional '--echoTE' 'echoTE' '@ delimited list of numbers' "TE for each
+
+opts_AddOptional '--wmprojabs' 'WMProjAbs' 'number' "FreeSurfer wm-proj-abs value" "2"
+
+opts_AddOptional '--runmode' 'RunMode' '1,2,3' "Which stages to run: 1=all, 2=skip GDC+MC, 3=OneStepResample+Intensity only" "1"
```
@takuya-hayashi @TakJim will discuss whether RunModes will be used in this pipeline.

glasserm replied:
I don't like the changes to the multi-echo code.

glasserm commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -420,29 +449,124 @@ if (( ! IsLongitudinal )); then
+				${FSLDIR}/bin/applywarp --rel --interp=spline -i ${WD}/Jacobian.nii.gz -r ${T1wImage} --premat="${WD}/${ScoutInputFileSE}_undistorted2T1w_init.mat" -o ${WD}/Jacobian2T1w.nii.gz
+
+				# 1-step resample from input (gdc) scout - NOTE: no longer includes jacobian correction, if specified
+				# Use SE filed magnitude for input of FS-BBR registration
```
field

TakJim replied:
Fixed it in 065e2207.

glasserm commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -731,6 +1019,8 @@ then
     fi
+else
+    
 fi
```
@TakJim to remove.

coalsont commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -573,79 +697,245 @@ if (( ! IsLongitudinal )); then
+                vnum=$(${FSLDIR}/bin/zeropad $((VolumeNumber + 1)) 2)
+                PhaseVol=One
+            fi
+            #imcp ${WD}/FieldMap/SBRef2Warpfield ${WD}/FieldMap/SBRef2Phase${PhaseVol}_gdc_initI 
```
Remove this commented line, .mat is not an image extension, so it wouldn't have worked.
```suggestion
```

coalsont commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -573,79 +697,245 @@ if (( ! IsLongitudinal )); then
+
+            # 3rd registration of distorted SBRef-to-SEField with FS-BBR
+            if [ $BBR = "T1w" ]; then
+                log_Msg "registeration of distorted T1w-contrast SBRef to distorted T1w with T1w-BBR"
```
```suggestion
                log_Msg "registration of distorted T1w-contrast SBRef to distorted T1w with T1w-BBR"
```

coalsont commented on fMRIVolume/scripts/MotionCorrection.sh:
```diff
@@ -113,6 +135,39 @@ case "$MotionCorrectionType" in
+        outputOrig=${WorkingDirectory}/${OutputfMRIBasenameOrig}
+        mkdir -p $outputOrig
+        pi=$(echo "scale=10; 4*a(1)" | bc -l)
+        for i in `ls ${output}/MAT_????.mat` ; do
```
Don't use `ls`, globbing works as-is.  Also, do we need to think about 10,000+ TRs in a single run?
```suggestion
        for i in "$output"/MAT_????.mat ; do
```

TakJim committed:
Simplify orientation check in fMRIVolume pipeline

TakJim committed:
Add .sh extension and fix variable style in CorrectVolumeOrientation calls

TakJim committed:
Use grep -A 3 instead of -3 for sform extraction

TakJim committed:
Remove commented imcp line

TakJim committed:
Fix typo: registeration -> registration

TakJim committed:
Use arithmetic test for UseJacobian

TakJim committed:
Use glob instead of ls for MAT file iteration

TakJim committed:
Rewrite fMRIVolume NHP batch script based on human batch and PreFreeSurferPipelineBatchNHP.sh

TakJim committed:
Fix review comments and pre-existing syntax bugs in fMRIVolume scripts

- Update jacobian comment: "handled below, and is optional" (r2772048473)
- Fix typo "filed" -> "field" (r2772049269)
- Remove empty else clause (r2772057701)
- Invert Human check to != Human pattern in OneStepResampling.sh (r2772060840)
- Add --species argument to DistortionCorrection script (r2776907266)
- Derive betspecieslabel from SPECIES when not set (r2776912100)
- Pass --species=${SPECIES} from pipeline to DistortionCorrection call
- Fix pre-existing syntax: [ && ] -> [[ && ]], missing space before ],
  missing fi for DistortionCorrection != SPIN_ECHO block

TakJim commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -420,29 +449,124 @@ if (( ! IsLongitudinal )); then
+				${FSLDIR}/bin/convertwarp --relout --rel -r ${T1wImage} --warp1=${WD}/WarpField.nii.gz --postmat="${WD}/${ScoutInputFileSE}_undistorted2T1w_init.mat" -o ${WD}/${ScoutInputFileSE}_undistorted2T1w_init_warp
+				${FSLDIR}/bin/applywarp --rel --interp=spline -i ${WD}/Jacobian.nii.gz -r ${T1wImage} --premat="${WD}/${ScoutInputFileSE}_undistorted2T1w_init.mat" -o ${WD}/Jacobian2T1w.nii.gz
+
+				# 1-step resample from input (gdc) scout - NOTE: no longer includes jacobian correction, if specified
```
Fixed it in 065e2207.

TakJim replied:
Updated the comment to clarify: jacobian modulation is handled below and is optional (33973fa4).

TakJim replied:
Deferred — after later edits, the target line in the `NONE` mode is no longer clearly identifiable. Could you point to the exact line you want the comment removed from in the current HEAD?

coalsont replied:
This was resolved already, ignore.

TakJim commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -420,29 +449,124 @@ if (( ! IsLongitudinal )); then
+				${FSLDIR}/bin/flirt -in ${WD}/${ScoutInputFileSE}_undistorted2T1w_initI -ref ${WD}/${T1wBrainImageFile} -omat ${WD}/${ScoutInputFileSE}_undistorted2T1w_initII.mat -interp spline -dof 6 -nosearch -o ${WD}/${ScoutInputFileSE}_undistorted2T1w_initII
+				${FSLDIR}/bin/fslmaths ${WD}/${ScoutInputFileSE}_undistorted2T1w_initII -mas ${WD}/${T1wBrainImageFile}_mask_dil ${WD}/${ScoutInputFileSE}_undistorted2T1w_initII
+
+				# fine tuning with FSL-BBR between fieldmap and T1w_acpc
```
Good point. Will add a clarifying comment to the `--bbr` option description noting that for NHP, it uses FSL FLIRT-based BBR (brain boundary or wmseg) rather than FreeSurfer surface-based BBR, and that it only applies when `--species` is not Human.

glasserm commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,479 @@
+#   NONE: no boundary-based registration
+#   T1w:  T1w-based BBR (e.g. for MION fMRI)
+#   T2w:  T2w-based BBR (default)
+BBR="T2w"
```
Change BBR variable to BBRContrast

glasserm commented on fMRIVolume/GenericfMRIVolumeMotionCorrectXRunsNHP.sh:
```diff
@@ -0,0 +1,314 @@
+#! /bin/bash 
```
This needs to be deleted.

coalsont commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,479 @@
+        PhaseInputName=""
+    fi
+
+    # Parse @-delimited session variables from hcppipe_conf.txt
```
Why are we expecting `hcppipe_conf.txt` to have an extra `@` on the front and/or end of every list?

TakJim commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -826,13 +851,57 @@ if (( ! IsLongitudinal )); then
+            initmat=""
+        fi
+        #NOTE: don't quote ${initmat}, we don't want to pass an empty string
+        "$GlobalScripts"/CorrectVolumeOrientation.sh --in="$fMRIFolder"/"$ScoutName"_gdc.nii.gz --out="$fMRIFolder"/"$ScoutName"_gdc --tposition="$TruePatientPosition" --sposition="$ScannerPatientPosition" --omat=TRUE ${initmat}
```
Rebased onto master (HEAD: 33973fa).

TakJim committed:
Fix simple review issues in fMRIVolume scripts

- Fix ${10}/${11} positional parameter expansion in MotionCorrection.sh
- Fix IsLongitudinal condition that always evaluated False
- Add --species option to OneStepResampling.sh
- Delete unused GenericfMRIVolumeMotionCorrectXRunsNHP.sh

TakJim committed:
Implement --runmode in fMRIVolume pipeline

glasserm commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -142,6 +163,18 @@ fi
+        Macaque*|Rhesus*|Cyno*) betspecieslabel=1 ;;
+        Marmoset)       betspecieslabel=0 ;;
+        NightMonkey)    betspecieslabel=0 ;;
+        *)              betspecieslabel=1 ;;
```
Fix the betspecieslabel codes for each species.

TakJim replied:
Fixed in e1dedd2e87407614463221459a7ab11e22d4e03f.

glasserm commented:
@coalsont We need to rebase this PR with the PreFreeSurfer merged PR so that all of this PR's dependencies are present. @mmilch01 has more details on this issue.

TakJim committed:
Fix review comments on PR #372

- Rename --bbr/BBR to --bbr-contrast/BBRContrast for clarity
- Fix jacobian comment in human code path, remove in NONE path
- Use (( ! IsLongitudinal )) syntax in DistortionCorrection script
- Restructure OneStepResampling.sh: full else for human/nonhuman
- Revert multi-echo code changes: restore else clause, remove duplicate

TakJim committed:
Correct handling of betspecieslabel

mmilch01 commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
This command would currently fail because tcsEchoesGdc, etc. are being defined in the next block. I'll wait for this to be fixed before continuing testing the human stream.

TakJim commented on fMRIVolume/GenericfMRIVolumeProcessingPipelineNHP.sh:
```diff
@@ -0,0 +1,566 @@
+	                                             DistortionCorrectionEPI2T1wReg;OneStepResampleIntensityNormalize;
+elif [[ $RunMode = 3 ]] ; then
+	                                                                            OneStepResampleIntensityNormalize;
+fi
```
The NHP-specific file has been deleted (05f7f646) and RunMode has been reimplemented in the unified GenericfMRIVolumeProcessingPipeline.sh using case fall-through style (0a44dcf9).

TakJim committed:
Add clarifying comments in OneStepResampling.sh per review

TakJim commented on fMRIVolume/scripts/OneStepResampling.sh:
```diff
@@ -162,13 +163,18 @@ NumFrames=$(${FSLDIR}/bin/fslval ${InputfMRI} dim4)
 #   NB: don't use FLIRT to do spline interpolation with -applyisoxfm for the
 #       2mm and 1mm cases because it doesn't know the peculiarities of the
 #       MNI template FOVs
```
Fixed in d53fe020.

TakJim committed:
Restore echo name array initialization and loop assignments

Complete the reversion of echo array changes that was flagged as
incomplete in PR #372 review comments r2853580713 and r2853593232.

TakJim commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -851,11 +941,6 @@ if [[ ${nEcho} -gt 1 ]]; then
-    sctEchoesOrig[0]="${OrigScoutName}"
-    tcsEchoesGdc[0]="${NameOffMRI}_gdc"
-    sctEchoesGdc[0]="${ScoutName}_gdc"
 fi
```
Fixed in fbdca99b.

TakJim committed:
Indent code inside RunMode conditionals for clarity

TakJim commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -826,22 +877,61 @@ if (( ! IsLongitudinal )); then
-        tcsEchoesOrig[iEcho]="${OrigTCSName}_E$(printf "%02d" "$iEcho")"
-        tcsEchoesGdc[iEcho]="${NameOffMRI}_gdc_E$(printf "%02d" "$iEcho")" # Is only first echo needed for the gdc tcs?
-        sctEchoesOrig[iEcho]="${OrigScoutName}_E$(printf "%02d" "$iEcho")"
-        sctEchoesGdc[iEcho]="${ScoutName}_gdc_E$(printf "%02d" "$iEcho")"
```
Fixed in fbdca99b.

TakJim commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -886,17 +985,21 @@ if (( IsLongitudinal )); then
+if [[ $nEcho -gt 1 ]] ; then
+    EchoDir="${fMRIFolder}/MultiEcho"
+    mkdir -p "$EchoDir"
 fi
```
This was intentionally moved outside the `if (( IsLongitudinal ))` block. With `--runmode`, Stage 1 may be skipped (RunMode >= 2), so the EchoDir setup inside Stage 1 (non-longitudinal path) wouldn't run. Moving it here ensures EchoDir is set for both longitudinal and non-longitudinal modes regardless of RunMode. The `fMRIFolder` variable has already been reassigned to the longitudinal path at this point if applicable, so EchoDir resolves correctly in both cases.

glasserm replied:
Revert https://github.com/Washington-University/HCPpipelines/commit/8eb67af94f28d340f5e380049e68eabd489181f7

TakJim commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -674,6 +715,12 @@ ResultsFolderCross="$AtlasSpaceFolder"/"$ResultsFolder"/"$NameOffMRI"
+if [ "$RunMode" -lt 2 ] ; then
+
 if (( ! IsLongitudinal )); then
     mkdir -p ${T1wFolder}/Results/${NameOffMRI}
```
Fixed in 23fcf87b.

TakJim committed:
Restore EchoDir setup inside IsLongitudinal block per review

TakJim committed:
Revert "Restore EchoDir setup inside IsLongitudinal block per review"

This reverts commit 8eb67af94f28d340f5e380049e68eabd489181f7.

mmilch01 commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,479 @@
+# fMRI name filter: leave empty to process all fMRI runs in each session
+# Set to a specific fMRI name (without extension) to process only that run
+fmriname=""
+
```
```suggestion
#structural image resolution
StructRes=0.5
```

mmilch01 replied:
Updated the suggestion.

TakJim replied:
Fixed in 37559893.

TakJim committed:
Simplify comment for NHP branch in OneStepResampling.sh

TakJim committed:
Update BatchNHP: add StructRes/isFLAIR, fix SetUpSPECIES args, remove @ parsing

- Add StructRes, isFLAIR, isT1wDivFLAIR variables
- Pass --structres to SetUpSPECIES.sh, pass args to SetUpFSNHP.sh
- Remove unnecessary @-delimiter stripping from session variables

TakJim commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,479 @@
+#   T1w:  T1w-based BBR (e.g. for MION fMRI)
+#   T2w:  T2w-based BBR (default)
+BBRContrast="T2w"
+
```
Fixed it in 17d2e6a7. Added `isFLAIR=0` and `isT1wDivFLAIR=1`.

TakJim commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,479 @@
+
+# Set up pipeline environment variables and software
+source "$EnvironmentScript"
+
```
Agreed with Tim — `--structres` is mandatory, so skipped the conditional. Passing `--structres="$StructRes"` directly in 17d2e6a7.

glasserm commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,487 @@
+
+# Set up species-specific environment variables
+source "$HCPPIPEDIR"/Examples/Scripts/SetUpSPECIES.sh --species="$SPECIES" --structres="$StructRes"
+source "$HCPPIPEDIR"/FreeSurfer/custom/SetUpFSNHP.sh "$SPECIES" "$isFLAIR" "$isT1wDivFLAIR"
```
```suggestion
source "$HCPPIPEDIR"/FreeSurfer/custom/SetUpFSNHP.sh "$SPECIES" "0" "0"
```

takuya-hayashi replied:
'WMProjAbs' is used to optimize FS BBR for each species.

coalsont replied:
It isn't clear to me that we would expect the scout and structural BBR parameters to always be exactly the same.  Diffusion appears to have its own setting for this in SetUpSPECIES (so, that isn't in `FreeSurfer/custom/SetUpFSNHP.sh`).

TakJim replied:
Fixed it in 6dc8e3de.

mmilch01 replied:
Put 'fMRI_WMProjAbs' with the same function in SetUpSPECIES.sh and use it everywhere outside of FreeSurferNHP. Do not source SetUpFSNHP.sh outside of FreeSurferNHP.
Put a comment in the batch file describing the option to override the value from the SetUpSPECIES.sh.

TakJim replied:
Fixed in 0aa23591 (SetUpSPECIES) and 57dff3e1 (batch).

- Added `FuncWMProjAbs` to `SetUpSPECIES.sh` per species (Human=2, Chimp=1, Macaque=0.7, Marmoset=0.2, NightMonkey=0.3; carried over from the existing `SetUpFSNHP.sh` values tuned by Takuya and team).
- Removed the `SetUpFSNHP.sh` source from the batch file.
- Renamed `$WMProjAbs` → `$FuncWMProjAbs` in the batch. The sub-script CLI option `--wmprojabs` is kept as-is since it matches FreeSurfer's `--wm-proj-abs` naming convention.
- Updated the inline comment block in the batch file to describe how to override `FuncWMProjAbs` from the `SetUpSPECIES.sh` default.

mmilch01 commented:
FinalFMRIResolution is not set in SetUpSpecies.sh for:

Macaque,
Cyno,
Rhesus,
Snow.

This would break processing for these species. Please add appropriate initializations to SetUpSpecies.sh.

mmilch01 commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,487 @@
+# Override here only if you need non-default values for your dataset.
+# Example values for Macaque:
+#   FinalFMRIResolution="1.25"
+#   TopUpConfig="${HCPPIPEDIR_Config}/b02b0.cnf"
```
```suggestion
#   TopupConfig="${HCPPIPEDIR_Config}/b02b0.cnf"
```
Make consistent with SetUpSpecies.sh

TakJim replied:
Fixed it in c4a83bd6.

mmilch01 commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,487 @@
+                --fmrires="$FinalFMRIResolution" \
+                --dcmethod="$DistortionCorrection" \
+                --gdcoeffs="$GradientDistortionCoeffs" \
+                --topupconfig="$TopUpConfig" \
```
```suggestion
                --topupconfig="$TopupConfig" \
```
Make consistent with SetUpSpecies.sh

TakJim replied:
Fixed it in c4a83bd6.

mmilch01 commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,487 @@
+                --fmrires=$FinalFMRIResolution \
+                --dcmethod=$DistortionCorrection \
+                --gdcoeffs=$GradientDistortionCoeffs \
+                --topupconfig=$TopUpConfig \
```
```suggestion
                --topupconfig=$TopupConfig \
```
Make consistent with SetUpSpecies.sh

TakJim replied:
Fixed it in c4a83bd6.

coalsont commented:
> FinalFMRIResolution is not set in SetUpSpecies.sh for:
> 
> Macaque, Cyno, Rhesus, Snow.
> 
> This would break processing for these species. Please add appropriate initializations to SetUpSpecies.sh.

I don't think that input image resolutions belong in SetUpSpecies.sh (and I think `FinalfMRIResolution` is just for volume timeseries, not the grayordinates space).  While the range of usable resolutions is different per species, I don't know that we should be expecting users to use one exact resolution in advance.  I think this should only be set in the Batch directly.

TakJim committed:
BatchNHP: rename TopUpConfig to TopupConfig for consistency with SetUpSPECIES

TakJim committed:
DistortionCorrection: remove unused SBRef2SEFieldByT1w block

TakJim commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -573,79 +721,246 @@ if (( ! IsLongitudinal )); then
+            ${FSLDIR}/bin/applywarp --rel --interp=spline -i ${WD}/FieldMap/SBRef.nii.gz -r ${WD}/FieldMap/SBRef.nii.gz -w ${WD}/WarpField -o ${WD}/FieldMap/SBRef_dc.nii.gz
+            ${FSLDIR}/bin/fslmaths ${WD}/FieldMap/SBRef_dc -mul ${WD}/FieldMap/Jacobian ${WD}/FieldMap/SBRef_dc_jac
+        fi  
+        ## Finish SBRef2SEFieldByT1w
```
Fixed it in 18aebd95. Removed lines 769-851 (the entire SBRef2SEFieldByT1w block) per TH's suggestion.

TakJim commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -573,79 +721,246 @@ if (( ! IsLongitudinal )); then
+        else # SPIN_ECHO_METHOD_OPT
+		    ## SBRef2SEFieldByT1w - tune up registration betweeen SBRef and SEFieldmap by way of T1w. This will update SBRef2Warpfield.mat by performing 3-step, 1) simulate distorted T1w based on topup, 2) BBR registration btw distorted SBRef and distorted T1w volume, and 3) BBR registration btw undistorted SBRef and T1w_acpc. This step may be effective if SBRef to SEFieldmap registration (in TopupPrep ...
+			SBRef2SEFieldByT1=FALSE  # TRUE or FALSE
+			
```
Fixed it in 18aebd95.

coalsont commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,487 @@
+#
+#### fMRIVolume-relevant variables
+#FinalFMRIResolution="2"     # Target final resolution of fMRI data in mm
+#TopupConfig="${HCPPIPEDIR_Config}/b02b0.cnf"  # Config for topup or "NONE" if not used
```
This looks like an old suggestion from before the "make consistent with SetUpSPECIES" fixes, I am marking it as resolved.

TakJim replied:
Skipped this one: `SetUpSPECIES.sh` uses `TopupConfig` (lowercase `up`), so changing to `TopUpConfig` here would diverge. Could you confirm the intended casing?

coalsont replied:
Case is a minor style choice (as long as it doesn't actually result in two variables in a script).  Ignore this.

mmilch01 commented:
> > FinalFMRIResolution is not set in SetUpSpecies.sh for:
> > Macaque, Cyno, Rhesus, Snow.
> > This would break processing for these species. Please add appropriate initializations to SetUpSpecies.sh.
> 
> I don't think that input image resolutions belong in SetUpSpecies.sh (and I think `FinalfMRIResolution` is just for volume timeseries, not the grayordinates space). While the range of usable resolutions is different per species, I don't know that we should be expecting users to use one exact resolution in advance. I think this should only be set in the Batch directly.

In this case for consistency, all definitions of  FinalFMRIResolution should be removed from SetUpSpecies.sh.

TakJim committed:
Fix StructRes default to 0.5 in BatchNHP

TakJim commented:
Rebased onto current master (`a2f3e5f3`). 38 commits, no conflicts. Diff contains only fMRIVolume-related files (no master changes leaked).

TakJim commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -420,29 +474,124 @@ if (( ! IsLongitudinal )); then
+				--scannerpatientposition=${ScannerPatientPosition} \
+                --truepatientposition=${TruePatientPosition} \
+                --initworldmat=${InitWorldMat} \
                 --usejacobian=${UseJacobian}
```
Done. Rebased onto current master (`a2f3e5f3`).

TakJim committed:
Fix indentation inside else SPIN_ECHO_METHOD_OPT block

TakJim commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -573,79 +721,246 @@ if (( ! IsLongitudinal )); then
+        fi  
+        ## Finish SBRef2SEFieldByT1w
+
+        cp ${WD}/${ScoutInputFileSE}_undistorted2T1w_init.mat ${WD}/${ScoutInputFileGE}_undistorted2T1w_init.mat
```
Fixed indentation in c6667715.

TakJim committed:
Move Human-only code into conditional blocks in DistortionCorrection

- Move Jacobian/registration/warpfield/SEBASED code (was shared) into
  the else # for Human block in TOPUP case
- Move UseJacobian+BiasField block into else # Human block in BBR section
- Delete old NHP stub log messages replaced by new NHP processing

TakJim commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
Applied in 5ea652d7. Moved lines 600-659 into the else # for Human block.

TakJim commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
Deleted and moved to the else # Human block in 5ea652d7.

TakJim commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -573,79 +721,246 @@ if (( ! IsLongitudinal )); then
+	    log_Msg "${FreeSurferSubjectFolder}/${FreeSurferSubjectID}_1mm does not exist. FreeSurferNHP.sh was not used."
+
+		# Run Normally
+        log_Msg "Run Normally" 
```
Applied in 5ea652d7. Moved UseJacobian block here and deleted the old log messages.

mmilch01 commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,487 @@
+# These are set by SetUpSPECIES.sh (sourced after EnvironmentScript).
+# Override here only if you need non-default values for your dataset.
+# Example values for Macaque:
+#   FinalFMRIResolution="1.25"
```
```suggestion
#   FinalfMRIResolution="1.25"
```
change to match SetUpSpecies.sh

TakJim replied:
Fixed it in 6dc8e3de.

mmilch01 commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,487 @@
+#CorticalScaleFactor="1"     # Cortical scale factor
+#
+#### fMRIVolume-relevant variables
+#FinalFMRIResolution="2"     # Target final resolution of fMRI data in mm
```
```suggestion
#FinalfMRIResolution="2"     # Target final resolution of fMRI data in mm
```
change to match SetUpSpecies.sh

TakJim replied:
Fixed it in 6dc8e3de.

mmilch01 commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,487 @@
+                --echospacing="$DwellTime" \
+                --echodiff="$DeltaTE" \
+                --unwarpdir="$UnwarpDir" \
+                --fmrires="$FinalFMRIResolution" \
```
```suggestion
                --fmrires="$FinalfMRIResolution" \
```
change to match SetUpSpecies.sh

TakJim replied:
Fixed it in 6dc8e3de.

mmilch01 commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,487 @@
+                --echospacing=$DwellTime \
+                --echodiff=$DeltaTE \
+                --unwarpdir=$UnwarpDir \
+                --fmrires=$FinalFMRIResolution \
```
```suggestion
                --fmrires=$FinalfMRIResolution \
```
change to match SetUpSpecies.sh

TakJim replied:
Fixed it in 6dc8e3de.

mmilch01 commented on fMRIVolume/scripts/MotionCorrection.sh:
```diff
@@ -92,10 +95,29 @@ verbose_echo "     OutputMotionMatrixFolder: ${OutputMotionMatrixFolder}"
+    # Scaling brain size to adapt to size dependency of mcflirt - TH 2024
+    if [[ ! -z "$BrainScaleFactor" && ! $(echo "$BrainScaleFactor == 1" | bc) = 1 ]] ; then
+        log_Msg "Scaling brain with a factor = $BrainScaleFactor"
+        ${HCPPIPEDIR_Global}/ScaleVolume.sh ${Scout}.nii.gz ${BrainScaleFactor} ${WorkingDirectory}/scout_scale.nii.gz ${WorkingDirectory}/scale.world.mat
```
```suggestion
        ${HCPPIPEDIR_Global}/ScaleVolumeNHP.sh ${Scout}.nii.gz ${BrainScaleFactor} ${WorkingDirectory}/scout_scale.nii.gz ${WorkingDirectory}/scale.world.mat
```

TakJim replied:
Fixed it in 87f0cdde.

TakJim committed:
Apply review suggestions in GenericfMRIVolumeProcessingPipelineBatchNHP.sh

- Rename FinalFMRIResolution -> FinalfMRIResolution to match SetUpSPECIES.sh
- Hardcode SetUpFSNHP.sh args and remove isFLAIR/isT1wDivFLAIR definitions

TakJim committed:
Rename ScaleVolume.sh -> ScaleVolumeNHP.sh in MotionCorrection.sh

TakJim commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,487 @@
+isFLAIR=0
+
+#Whether T2 image is T1w divided by FLAIR, 0 or 1
+isT1wDivFLAIR=1
```
Fixed it in 6dc8e3de.

TakJim commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
This code is already placed inside the `else # for Human` conditional (line 588 `else # for Human` ... `fi #end Human conditional` at the end of the human-only block). Please let me know if a further restructuring is needed.

TakJim commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,479 @@
+
+# Set up species-specific environment variables
+source "$HCPPIPEDIR"/Examples/Scripts/SetUpSPECIES.sh --species="$SPECIES"
+source "$HCPPIPEDIR"/FreeSurfer/custom/SetUpFSNHP.sh
```
Already applied in current HEAD (line 132). Note that `SetUpFSNHP.sh` args were subsequently changed to `"0" "0"` per @glasserm's follow-up suggestion.

TakJim committed:
SetUpSPECIES: add FuncWMProjAbs per species

TakJim committed:
fMRI batch: stop sourcing SetUpFSNHP, use FuncWMProjAbs from SetUpSPECIES

- Remove cross-pipeline source of FreeSurfer/custom/SetUpFSNHP.sh.
  FuncWMProjAbs is now defined per species in SetUpSPECIES.sh.
- Rename WMProjAbs to FuncWMProjAbs in batch (CLI option --wmprojabs
  to sub-script is kept unchanged to match FreeSurfer --wm-proj-abs).
- Update inline comment describing how to override FuncWMProjAbs
  from the default set by SetUpSPECIES.sh.

Addresses review comment by mmilch01 on PR #372.

coalsont commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,483 @@
+#   TopupConfig="${HCPPIPEDIR_Config}/b02b0.cnf"
+#   FuncWMProjAbs="1"
+#   BrainScaleFactor="0.36"
+#   betspecieslabel="1"
```
If a user uncomments some of these, and then SetUpSPECIES changes the values back (below), the user will probably be confused.  Let's not put any `#   FinalfMRIResolution=...`-type patterns for variables that SetUpSPECIES sets above the SetUpSPECIES `source` line.

There is already some explanation of some of these type of settings below the SetUpSPECIES `source` line (134), so this should be merged into the lower part (rather than have multiple sections that basically say the same thing). And also set the variables appropriately for macaque.

TakJim replied:
Fixed in b594c51c — removed the pre-source example block and merged the documentation into the post-source block, with values updated to macaque to match the default SPECIES=Macaque.

glasserm commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,483 @@
+        TopupPositive2=""
+        TopupNegative2=""
+        Gradient=""
+        MagnitudeInputName=""
+        PhaseInputName=""
+    fi
```
```suggestion
    else
      echo "${SCRIPT_NAME}: hcppipe_conf.txt not found for ${Subject}, using inline defaults"
    fi
    # Set these variables manually when hcppipe_conf.txt is not available.
    # Sessions are separated by @ delimiters. Within each session, runs are space-delimited.
    # Example for a single session with two fMRI runs:
    #   Tasklist="rest1 task1"
    # Example for two sessions (separated by @):    
    #   Tasklist="rest1 task1@rest2 task2"    
    #   Taskreflist="rest1_SBRef task1_SBRef"
    #   TopupPositive="SE_AP SE_AP"
    #   TopupNegative="SE_PA SE_PA"
    #   PhaseEncodinglist="y y"
    #   Fmriconcatlist=""    
    #   DwellTime="0.00058"
    #   InitWorldMat=""    
    #   TruePatientPosition="HFS"
    #   ScannerPatientPosition="HFS"
    #   TopupPositive2=""
    #   TopupNegative2=""
    #   Gradient=""
    #   MagnitudeInputName=""
    #   PhaseInputName=""
```

TakJim replied:
Fixed in c7a279e5 — applied your range suggestion verbatim.

TakJim committed:
GenericfMRIVolumeProcessingPipelineBatchNHP.sh: drop inline defaults

Per Matt suggestion r3127051582 (range L206-236): when
hcppipe_conf.txt is absent, only emit the warning and let the
user set the variables manually in their batch wrapper. Move
the example variable list out of the else block to a comment
block following the if/fi, so the example is visible in either
branch and not tied to the empty-default behavior.

Net: -29/+21 lines.

TakJim committed:
fMRI batch: consolidate SetUpSPECIES override docs, use macaque values

Per Tim r3126965922: the pre-source example block (above the
SetUpSPECIES.sh source line) duplicated the post-source block
and would silently be overridden by SetUpSPECIES, confusing
users who thought they were setting overrides.

- Remove the pre-source example block (L100-108).
- Update the post-source description to use Macaque values
  (matching the default SPECIES=Macaque) so users see the
  values that actually apply unless overridden.

Net: -31/+18 lines.

coalsont commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,462 @@
+# Set up pipeline environment variables and software
+source "$EnvironmentScript"
+
+# Set up species-specific environment variables
+source "$HCPPIPEDIR"/Examples/Scripts/SetUpSPECIES.sh --species="$SPECIES" --structres="$StructRes"
```
~Include the hack from getting the `newopts` dependencies to behave well in a "`source` in example batch" context:~

`<suggestion removed>`

coalsont replied:
Actually, I can fix this specific thing in `SetUpSPECIES` itself...

coalsont commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,462 @@
+EnvironmentScript="${HOME}/projects/Pipelines/Examples/Scripts/SetUpHCPPipeline.sh"
+
+# Species label (Macaque, MacaqueCyno, MacaqueRhesus, Marmoset, NightMonkey, Chimp, Human)
+SPECIES="Macaque"
```
"Macaque" will error out in `SetUpSPECIES.sh`, because it isn't specific enough.  The combination of `betfraction="0.3"` and `betcenter="48,56,51"` in the example lines suggests the example is for "MacaqueSnow" or "SnowMacaque" (which isn't mentioned in the list of options, and neither is "MacaqueMac30BS", though maybe that was excluded on purpose).  EDIT: "MacaqueRhesus" was the more likely intent.

```suggestion
# Species label (MacaqueCyno, MacaqueRhesus, MacaqueSnow, Marmoset, NightMonkey, Chimp, Human)
SPECIES="MacaqueRhesus"
```

takuya-hayashi replied:
Thanks. @TakJim Please update as Tim suggested.

TakJim replied:
Applied in 06f23706 — switched default to `MacaqueRhesus` and updated the species list in the comment.

coalsont commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,462 @@
+# Set up species-specific environment variables
+source "$HCPPIPEDIR"/Examples/Scripts/SetUpSPECIES.sh --species="$SPECIES" --structres="$StructRes"
+
+# The script ${HCPPIPEDIR}/Examples/Scripts/SetUpSPECIES.sh defines per-species
+# values. The example values below correspond to Macaque (matching the default
+# SPECIES="Macaque" above). Uncomment and edit any line to override the
+# SetUpSPECIES.sh default for your dataset.
```
Match a valid `SPECIES` setting.  ~Assuming Snow Macaque, as above~:
```suggestion
# The script ${HCPPIPEDIR}/Examples/Scripts/SetUpSPECIES.sh defines per-species
# values. The example values below correspond to Rhesus Macaque (matching the default
# SPECIES="MacaqueRhesus" above). Uncomment and edit any line to override the
# SetUpSPECIES.sh default for your dataset.
```

TakJim replied:
Applied in 81584b5a.

coalsont commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,462 @@
+#### PreFreeSurfer-relevant variables (not used in fMRIVolume)
+#BrainSize="60"                # BrainSize in mm, distance between top of FOV and bottom of brain
+#betcenter="48,56,51"          # Comma separated voxel coordinates in T1wTemplate2mm
+#betradius="35"                # Brain radius for bet
```
MacaqueSnow doesn't set 35 here.  Alternatively, MacaqueRhesus sets `betfraction="0.2"`.

```suggestion
#betradius="40"                # Brain radius for bet
```

TakJim replied:
Applied in b1367e25 — `#betradius` example now matches the MacaqueRhesus default (40).

coalsont commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,462 @@
+#TopupConfig="${HCPPIPEDIR_Config}/b02b0_macaque_fMRI.cnf"  # Config for topup or "NONE" if not used
+#FuncWMProjAbs="0.7"           # fMRI wm-proj-abs for FreeSurferBBR
+#betspecieslabel="1"           # Species label for bet4animal (0=smallest, 4=largest brain)
+#betfraction="0.3"             # Fractional intensity threshold for bet
```
Match Rhesus setting:

```suggestion
#betfraction="0.2"             # Fractional intensity threshold for bet
```

TakJim replied:
Applied in 968a8c81 — `#betfraction` example now matches the MacaqueRhesus default (0.2).

coalsont commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,462 @@
+#FinalfMRIResolution="1.2"     # Target final resolution of fMRI data in mm
+#TopupConfig="${HCPPIPEDIR_Config}/b02b0_macaque_fMRI.cnf"  # Config for topup or "NONE" if not used
+#FuncWMProjAbs="0.7"           # fMRI wm-proj-abs for FreeSurferBBR
+#betspecieslabel="1"           # Species label for bet4animal (0=smallest, 4=largest brain)
```
This doesn't appear to be in SetUpSPECIES anymore, it seems to be handled from `SPECIES` automatically in `DistortionCorrection...` instead.

```suggestion
```

TakJim replied:
Removed in 64899d51 — agreed, `betspecieslabel` is derived from SPECIES inside `DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh` now, no example override needed here.

coalsont commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatchNHP.sh:
```diff
@@ -0,0 +1,462 @@
+#T1wTemplate, T1wTemplateBrain, T1wTemplate2mm, T2wTemplate, etc.
+
+# Export species-specific environment variables used by subscripts
+export betspecieslabel
```
Exporting things other than tool locations is something we don't want to rely on.  We already handle `betspecieslabel` internally.

```suggestion
```

TakJim replied:
Removed in 33c302e5.

TakJim committed:
fMRI batch: remove unnecessary 'export betspecieslabel'

Per coalsont review on PR #372 — betspecieslabel is now handled
internally from SPECIES, no need to export it from the batch script.

TakJim committed:
fMRI batch: align example betradius with MacaqueRhesus default (35 -> 40)

Per coalsont review on PR #372.

TakJim committed:
fMRI batch: align example betfraction with MacaqueRhesus default (0.3 -> 0.2)

Per coalsont review on PR #372.

TakJim committed:
fMRI batch: drop example betspecieslabel override

Per coalsont review on PR #372 — betspecieslabel is not in
SetUpSPECIES anymore; it is derived from SPECIES inside
DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh.

TakJim committed:
fMRI batch: update override-block comment to match MacaqueRhesus default

Per coalsont review on PR #372.

TakJim committed:
fMRI batch: switch default SPECIES from 'Macaque' to 'MacaqueRhesus'

Per coalsont review on PR #372 — 'Macaque' alone errors out in
SetUpSPECIES.sh because it isn't specific enough. Use MacaqueRhesus
as the example default and list valid species in the comment.

coalsont commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -245,8 +289,7 @@ case $DistortionCorrection in
 
-if (( ! IsLongitudinal )); then 
-
+if (( ! IsLongitudinal )) || [[ "$SPECIES" != "Human" ]]; then # - for NHP TH 2017-2024
```
This logic appears to implement "if not human, ignore longitudinal arguments and run non-longitudinal".  We may instead want to explicitly check for "NHP and longitudinal" above and error, if that isn't supported.

coalsont replied:
```suggestion
if (( ! IsLongitudinal )); then
```

TakJim replied:
Fixed it in b4174634.

glasserm commented:
Will need to rebase to merged master after FreeSurferNHP is merged and script name has been changed to ScalevolumeNHP.sh

coalsont commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -142,6 +163,29 @@ fi
+    betspecieslabel="4"
+    ;;
+  *)
+    betspecieslabel=""
+    log_Err_Abort "Invalid species: '$SPECIES'. Must be one of: Human, Macaque, Rhesus, Chimp, NightMonkey, Marmoset."
+    ;;
+esac
+
 if (( $IsLongitudinal )); then
     if [ ! -f "$T1wCross2LongXfm" ]; then
         log_Err_Abort "--t1w-cross2long-xfm must point to a valid file when --is-longitudinal is used"
```
```suggestion

if (( $IsLongitudinal )); then
    if [[ "$SPECIES" != "Human" ]]
    then
        log_Err_Abort "longitudinal mode is not supported for non-human species"
    fi
    if [ ! -f "$T1wCross2LongXfm" ]; then
        log_Err_Abort "--t1w-cross2long-xfm must point to a valid file when --is-longitudinal is used"
```

TakJim replied:
Fixed it in b4174634.

glasserm replied:
@TakJim the above seems like the wrong commit to reference for this.

coalsont replied:
The change was made in the PR branch, appears to be ab227eeaeb589f31cb41150938b814f8ba227cc4.

glasserm commented:
@TakJim can you finish these commits please?

TakJim committed:
Apply review suggestions: early abort for NHP+longitudinal in DistortionCorrection

- Add explicit error in longitudinal block when SPECIES != Human
- Simplify line 292 conditional since longitudinal+non-Human is now caught early

TakJim commented:
Rebased onto master (commit `ab227eea`). PR #364 (Unified NHP FreeSurferPipeline) is now merged, and the script reference (`ScaleVolumeNHP.sh`) was already in this branch (commit `87f0cdde`). Rebase resolved cleanly with no conflicts; PR diff remains scoped to the 6 fMRI files.

coalsont commented:
What is the status on testing this?

mmilch01 commented:
> What is the status on testing this?

Test on Barbosa completed successfully yesterday.
